pkg/vectorize/date_add/date_add.go

```diff
@@ -13,28 +13,55 @@
 // limitations under the License.
 package date_add
 
-import "github.com/matrixorigin/matrixone/pkg/container/types"
+import (
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+)
 
 var (
-	DateAdd     func([]types.Date, []int64, []int64, []types.Date) []types.Date
-	DatetimeAdd func([]types.Datetime, []int64, []int64, []types.Datetime) []types.Datetime
+	DateAdd       func([]types.Date, []int64, []int64, []types.Date) []types.Date
+	DatetimeAdd   func([]types.Datetime, []int64, []int64, []types.Datetime) []types.Datetime
+	DateStringAdd func(*types.Bytes, []int64, []int64, *nulls.Nulls, *types.Bytes) *types.Bytes
 )
 
 func init() {
 	DateAdd = dateAdd
 	DatetimeAdd = datetimeAdd
+	DateStringAdd = dateStringAdd
 }
 
 func dateAdd(xs []types.Date, ys []int64, zs []int64, rs []types.Date) []types.Date {
 	for i, d := range xs {
-		rs[i] = d.ToTime().AddInterval(ys[i], types.IntervalType(zs[i])).ToDate()
+		rs[i] = d.ToTime().AddInterval(ys[0], types.IntervalType(zs[0])).ToDate()
 	}
 	return rs
 }
 
 func datetimeAdd(xs []types.Datetime, ys []int64, zs []int64, rs []types.Datetime) []types.Datetime {
 	for i, d := range xs {
-		rs[i] = d.AddInterval(ys[i], types.IntervalType(zs[i]))
+		rs[i] = d.AddInterval(ys[0], types.IntervalType(zs[0]))
+	}
+	return rs
+}
+
+func dateStringAdd(xs *types.Bytes, ys []int64, zs []int64, ns *nulls.Nulls, rs *types.Bytes) *types.Bytes {
+	for i := range xs.Lengths {
+		str := string(xs.Get(int64(i)))
+		if types.UnitIsDayOrLarger(types.IntervalType(zs[0])) {
+			d, e := types.ParseDate(str)
+			if e == nil {
+				rs.AppendOnce([]byte(d.ToTime().AddInterval(ys[0], types.IntervalType(zs[0])).ToDate().String()))
+				continue
+			}
+		}
+		d, e := types.ParseDatetime(str)
+		if e != nil {
+			// set null
+			nulls.Add(ns, uint64(i))
+			rs.AppendOnce([]byte(""))
+			continue
+		}
+		rs.AppendOnce([]byte(d.AddInterval(ys[0], types.IntervalType(zs[0])).String()))
```

Comment:
is there a particular reason that you wanna return varchar rather the Datetime?

Reply:
the reason is when the table have three varchar, ("2022-01-01", "2022-01-01 00:00:00", "xxxx"), the result should be ("2022-01-01", "2022-01-01 00:00:00", null)
so we have two return types, date and datetime.